woocommerce/woocommerce-ios
# Order Details: show aggregated order items for a shipping label

Part of #2167 

Depends on https://github.com/woocommerce/woocommerce-ios/pull/3188

This PR aggregates order items from an array of shipping labels, order items, products, and product variations from an order. The logic is a bit complicated due to the interesting API behavior, I tried adding comments along the way but feel free to ask any questions or request any explanations.

## Changes

- Created `ShippingLabelOrderItemsAggregator` that aggregates order items from an array of shipping labels, order items, products, and product variations from an order. The availability of products and product variations depends on whether they have been synced before. Note that syncing for product variations associated with an order is saved for a separate subtask of #2167.
- In `OrderDetailsResultsControllers`, added a FRC for `ProductVariation`s associated with an order.
- In `OrderDetailsDataSource`, set `ShippingLabelOrderItemsAggregator` when shipping labels are set in `reloadSections`. Updated the number of order item rows from `ShippingLabelOrderItemsAggregator`'s `orderItems` size.
- In `ProductDetailsCellViewModel`, added `productImageURL: URL?` so that the image URL is available in `AggregateOrderItem` without a concrete `Product` (it could also be `ProductVariation`).
- In `OrderDetailsViewModel`, handled tapping on a shipping label order item (product).
- Updated `AggregateOrderItem` from a `class` to `struct` so that we can use the auto-generated `Equatable` implementation.
- Added unit tests for `ShippingLabelOrderItemsAggregator`.

## Testing

Prerequisites: the site has [WooCommerce Shipping & Tax plugin](https://docs.woocommerce.com/document/woocommerce-shipping-and-tax/) and there is an order with at least one non-refunded shipping label.

You can set up your store to use a test credit card with instructions in p91TBi-3xD-p2.

**Notes**: if the store has not fetched product variations before order details, you might see the variation cell showing price 0.0. There are separate subtasks to fix this.

- Go to the orders tab
- Tap on an order with at least one non-refunded shipping label --> after syncing, a shipping label package card should be shown "Package #" with order items (products/variations) that are included in this package with the correct quantity/price (feel free to compare with core or Android)
  - Note that if the product/variation hasn't been synced, only the product name is shown.

## Example screenshots

Notes for @Garance91540:
- Please only focus on the order item rows (the other rows might not have the latest design yet)
- I plan to hide the 0 quantity/price labels in the product/variation cell if the information isn't available yet in a separate subtask. There is also another subtask to sync product variations associated with an order.

before variations are synced | after variations are synced (from opening the variations from a variable product in the Products tab)
-- | --
![Simulator Screen Shot - iPhone 11 - 2020-11-24 at 14 30 42](https://user-images.githubusercontent.com/1945542/100059898-c44f3800-2e66-11eb-88ab-537a076ba80d.png) | ![Simulator Screen Shot - iPhone 11 - 2020-11-24 at 14 30 57](https://user-images.githubusercontent.com/1945542/100059913-c87b5580-2e66-11eb-93d7-adae68f4002b.png)


Update release notes:
